[feat] 홈 날씨 API

rafa-e1 opened the pull request:
## 📚 PR 요약
홈 날씨 API 구현

## 📝 작업 내용
- WeatherKit 추가
- 위치 접근 권한 요청
- 사용자 위치 기반으로 날씨 이미지 적용

## 📌 참고 사항


## 📷 Screenshot


## 📮 관련 이슈
- Resolved: #195

rafa-e1 committed:
[chore] 앱 최소 버전 업그레이드

- 15.0 -> 16.0

rafa-e1 committed:
[chore] WeatherKit 추가

rafa-e1 committed:
[feat] 날씨 및 위치 api 구현

rafa-e1 committed:
[chore] 위치 접근 권한 요청 추가

rafa-e1 committed:
[feat] 뷰 모델 적용

rafa-e1 committed:
[feat] 날씨 api 뷰 연결

minnnidev commented on BookTalk/BookTalk/Sources/Network/Home/LocationManager.swift:
```diff
@@ -0,0 +1,59 @@
+//
+//  LocationManager.swift
```
위치 관련 정보를 처리하는 모듈은 네트워크 작업이 아니기 때문에 Home에 `Manager` 폴더를 생성하거나 `Application` 폴더로 이동하면 더 적절하지 않을까라는 생각이 듭니다!

rafa-e1 replied:
넵 감사합니다 수정하도록 하겠습니다!

minnnidev commented on BookTalk/BookTalk/Sources/Network/Home/WeahterService.swift:
```diff
@@ -0,0 +1,30 @@
+
+        do {
+            let weather = try await weatherService.weather(for: location)
+            delegate?.didUpdateWeatherCondition(weather.currentWeather.condition)
```
```swift
    func fetchWeather(for location: CLLocation) async -> WeatherCondition? {
        let weatherService = WeatherService()

        do {
            let weather = try await weatherService.weather(for: location)
            return weather.currentWeather.condition
        } catch {
            print("DEBUG: Weather fetching error \(error)")
            return nil
        }
    }
```

직접 돌려보진 못했지만, `delegate` 패턴을 쓰지 않고 그냥 리턴값으로 주면 더 코드가 간결해질 것 같아요

rafa-e1 replied:
오 넵 감사합니다!

minnnidev commented on BookTalk/BookTalk/Sources/Presentation/Home/ViewModel/HomeViewModel.swift:
```diff
@@ -38,8 +46,85 @@ final class HomeViewModel {
+        }
+    }
+
+    func backgroundImage(for condition: WeatherCondition?) -> UIImage? {
```
뷰에 보여줄 데이터를 처리한다는 용도로 뷰모델에 해당 메서드를 작성하는 게 적합하긴 하지만, 개인적으로는 코드가 많이 길고 엄청나게 핵심적인 코드도 아니라는 생각이 들어서

1. ViewModel의 extension을 추가하여 밑쪽으로 내려준다
2. 헬퍼 클래스를 따로 만든다
3. `UIImage` 의 extension으로 작성한다 

등 여러 방안을 이용해서 조금의 분리가 있으면 더 가독성이 좋아질 것 같습니다! + 중복된 리턴값을 가지는 케이스는 함께 처리

rafa-e1 replied:
오오오 넵 감사합니다!

minnnidev commented on BookTalk/BookTalk/Sources/Presentation/Home/Cell/TableViewCell/SuggestionCell.swift:
```diff
@@ -17,23 +18,11 @@ final class SuggestionCell: BaseTableViewCell {
-        }
-
-        viewModel.send(action: .loadBackgroundImageView)
+        backgroundImageView.image = viewModel.backgroundImage(for: weatherCondition)
```
앗 이걸 왜 이제야 봤지 싶긴 한데, 개인적으로 저는 cell이 데이터 표시하는 역할을 중점으로 해야 한다고 생각해서요.

현재 cell은 뷰모델을 직접 호출하지 않고 그냥 이미지 자체를 바인딩하는 게 저는 좀 더 적절하다고 생각하는데, 어떻게 생각하시나요?

rafa-e1 replied:
아 저도 그게 코드 일관성도 지키고 역할상 맞는 것 같습니다! 수정하겠습니다!

rafa-e1 committed:
[move] LocationManager 파일 위치 변경

rafa-e1 committed:
[refactor] delegate 패턴 제거 후 리턴값으로 변경

rafa-e1 committed:
[refactor] SuggestionCell의 ViewModel 의존성 제거 및 HomeViewController로 로직 이동

rafa-e1 committed:
[refactor] UIImage extension을 통해 날씨 상태에 따른 이미지 처리 로직을 ViewModel에서 분리

- ViewModel에서 뷰 관련 처리 코드를 제거하여 역할 분리
- UIImage extension을 추가하여 날씨 상태에 맞는 이미지를 처리
- 중복된 리턴 값을 하나로 통합해 가독성 향상 및 코드 간결화
- View에서 더 간결하게 이미지를 바인딩하도록 수정
